### ISSUE TITLE
Implementing event listener with proxy fails in native image

### ISSUE THREAD

**Issue by mmoayyed** [spring-projects/spring-framework#29872]
## Environment

Using Spring Boot 3.0.2 with the following:

```
java version "17.0.5" 2022-10-18 LTS
Java(TM) SE Runtime Environment GraalVM EE 22.3.0 (build 17.0.5+9-LTS-jvmci-22.3-b07)
Java HotSpot(TM) 64-Bit Server VM GraalVM EE 22.3.0 (build 17.0.5+9-LTS-jvmci-22.3-b07, mixed mode, sharing)

------------------------------------------------------------
Gradle 7.5
------------------------------------------------------------

Build time:   2022-07-14 12:48:15 UTC
Revision:     c7db7b958189ad2b0c1472b6fe663e6d654a5103

Kotlin:       1.6.21
Groovy:       3.0.10
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          17.0.5 (Oracle Corporation 17.0.5+9-LTS-jvmci-22.3-b07)
OS:           Mac OS X 13.1 aarch64
```
## Problem

```
./build/native/nativeCompile/demo

 :: Spring Boot ::                (v3.0.2)

2023-01-20T15:37:34.264+04:00  WARN 29315 --- [           main] w.s.c.ServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'myEventListener': Instantiation of supplied bean failed
2023-01-20T15:37:34.264+04:00  INFO 29315 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
2023-01-20T15:37:34.265+04:00 ERROR 29315 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'myEventListener': Instantiation of supplied bean failed
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainInstanceFromSupplier(AbstractAutowireCapableBeanFactory.java:1236) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainFromSupplier(AbstractAutowireCapableBeanFactory.java:1210) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1157) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[demo:6.0.4]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:961) ~[demo:6.0.4]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:915) ~[demo:6.0.4]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:584) ~[demo:6.0.4]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[demo:3.0.2]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730) ~[demo:3.0.2]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:432) ~[demo:3.0.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308) ~[demo:3.0.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1302) ~[demo:3.0.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1291) ~[demo:3.0.2]
	at com.example.demo.DemoApplication.main(DemoApplication.java:38) ~[demo:na]
Caused by: com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface com.example.demo.MyListener] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89) ~[na:na]
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.proxy.DynamicProxySupport.getProxyClass(DynamicProxySupport.java:171) ~[na:na]
	at java.base@17.0.5/java.lang.reflect.Proxy.getProxyConstructor(Proxy.java:47) ~[demo:na]
	at java.base@17.0.5/java.lang.reflect.Proxy.newProxyInstance(Proxy.java:1037) ~[demo:na]
	at com.example.demo.MyConfiguration$MyInnerConfiguration.getProxy(MyConfiguration.java:34) ~[demo:na]
	at com.example.demo.MyConfiguration$MyInnerConfiguration.myEventListener(MyConfiguration.java:30) ~[demo:na]
	at com.example.demo.MyConfiguration__BeanDefinitions$MyInnerConfiguration__BeanDefinitions.lambda$getMyEventListenerInstanceSupplier$0(MyConfiguration__BeanDefinitions.java:41) ~[na:na]
	at org.springframework.util.function.ThrowingFunction.apply(ThrowingFunction.java:63) ~[demo:6.0.4]
	at org.springframework.util.function.ThrowingFunction.apply(ThrowingFunction.java:51) ~[demo:6.0.4]
	at org.springframework.beans.factory.aot.BeanInstanceSupplier.lambda$withGenerator$0(BeanInstanceSupplier.java:173) ~[na:na]
	at org.springframework.util.function.ThrowingBiFunction.apply(ThrowingBiFunction.java:68) ~[demo:6.0.4]
	at org.springframework.util.function.ThrowingBiFunction.apply(ThrowingBiFunction.java:54) ~[demo:6.0.4]
	at org.springframework.beans.factory.aot.BeanInstanceSupplier.lambda$get$2(BeanInstanceSupplier.java:208) ~[na:na]
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:59) ~[demo:6.0.4]
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:47) ~[demo:6.0.4]
	at org.springframework.beans.factory.aot.BeanInstanceSupplier.invokeBeanSupplier(BeanInstanceSupplier.java:220) ~[na:na]
	at org.springframework.beans.factory.aot.BeanInstanceSupplier.get(BeanInstanceSupplier.java:208) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainInstanceFromSupplier(AbstractAutowireCapableBeanFactory.java:1225) ~[demo:6.0.4]
```

The following proxy config is generated:

```json
[
  {
    "interfaces": [
      "com.example.demo.MyListener",
      "org.springframework.aop.SpringProxy",
      "org.springframework.aop.framework.Advised",
      "org.springframework.core.DecoratingProxy"
    ]
  }
]

```

## Steps to reproduce

- Here is a reproducer: [demo.zip](https://github.com/spring-projects/spring-boot/files/10465881/demo.zip)

- Unzip, and run `./gradlew clean build` and then `./gradlew nativeBuild`.
- Run the generated native application

**Comment by sbrannen** [comment]
> If I may, is this really not something that could/should be handled by the plugin?

If you invoke `Proxy.newProxyInstance(...)` within your own code, Spring is not aware of that, and you are responsible for registering the necessary runtime hints to support the proxy you are requesting.

Though I think the more important issue is figuring out why you are creating proxies of your listeners and why Spring is also creating proxies of your listeners.

Based on the reflection configuration you supplied, it appears that there are three (3) proxies being created for your `MyListener` instance.

```json
  {
    "interfaces": [
      "com.example.demo.MyListener"
    ]
  },
```

The above looks like it comes from a manual JDK proxy created based solely on the `MyListener` interface.

```json
  {
    "interfaces": [
      "com.example.demo.MyListener",
      "org.springframework.aop.SpringProxy",
      "org.springframework.aop.framework.Advised",
      "org.springframework.core.DecoratingProxy"
    ]
  },
```

The above looks like it comes from Spring AOP creating a JDK proxy based on the `MyListener` interface plus the three additional Spring interfaces used by Spring AOP.

```json
  {
    "interfaces": [
      "com.example.demo.MyListener",
      "java.io.Serializable",
      "org.springframework.aop.SpringProxy",
      "org.springframework.aop.framework.Advised",
      "org.springframework.core.DecoratingProxy"
    ]
  }
```

The above looks like it comes from Spring AOP creating a JDK proxy created based on the `MyListener` interface plus the three additional Spring interfaces used by Spring AOP, but this time an instance of the proxy created for one of the previous two scenarios was supplied -- since the proxy now implements `Serializable`, and IIRC every JDK proxy instance implements `Serializable`.

In other words -- without having debugged the code in a native image -- it appears that you have three levels of proxies: some created by you and some created by Spring.

**Comment by sdeleuze** [comment]
Re-reading the discussion here, I don't see an actionable item here. If the user project code is creating a proxy, it should registers related hints. If Spring Cloud is registering additional proxies for `@RefreshScope`, it should be its responsibility to register related hints.
